README
ESCPOS
======

Python library to manipulate ESC/POS Printers.

1. Dependencies
---------------

In order to start getting access to your printer, you must ensure
you have previously installed the following python modules:

  * pyusb (python-usb)
  * PIL (Python Image Library)

2. Description
--------------

Python ESC/POS is a library which lets the user have access to all
those printers handled by ESC/POS commands, as defined by Epson,
from a Python application.

The standard usage is send raw text to the printer, but in also
helps the user to enhance the experience with those printers by
facilitating the bar code printing in many different standards,
as well as manipulating images so they can be printed as brand
logo or any other usage images migh have.

Text can be aligned/justified and fonts can be changed by size,
type and weight.

Also, this module handles some hardware functionalities like, cut
paper, carrier return, printer reset and others concerned to the
carriage alignment.

3. Define your printer
----------------------

Before start create your Python ESC/POS printer instance, you must
see at your system for the printer parameters. This is done with
the 'lsusb' command.

First run the command to look for the "Vendor ID" and "Product ID",
then write down the values, these values are displayed just before
the name of the device with the following format:

  xxxx:xxxx

Example:
  Bus 002 Device 001: ID 1a2b:1a2b Device name

Write down the the values in question, then issue the following
command so you can get the "Interface" number and "End Point"

  lsusb -vvv -d xxxx:xxxx | grep iInterface
  lsusb -vvv -d xxxx:xxxx | grep bEndpointAddress | grep OUT

The first command will yields the "Interface" number that must
be handy to have and the second yields the "Output Endpoint"
address.

By default the "Interface" number is "0" and the "Output Endpoint"
address is "0x82",  if you have other values then you can define
with your instance.


4. Define your instance
-----------------------

The following example shows how to initialize the Epson TM-TI88IV
*** NOTE: Always finish the sequence with Epson.cut() otherwise
          you will endup with weird chars being printed.

  from escpos import *

  """ Seiko Epson Corp. Receipt Printer M129 Definitions (EPSON TM-T88IV) """
  Epson = escpos.Escpos(0x04b8,0x0202,0)
  Epson.text("Hello World")
  Epson.image("logo.gif")
  Epson.barcode
  Epson.barcode('1324354657687','EAN13',64,2,'','')
  Epson.cut()

5. Links
--------

Please visit project homepage at:
http://repo.bashlinux.com/projects/escpos.html

* Manuel F Martinez <manpaz@bashlinux.com>
* Dmitry Orlov <me@mosquito.su>


CHANGELOG
CHANGELOG

* 2014-03-28 - Version 1.0.3
- Add codepage settings
- Add close method
- Register to PYPI as python-escpos
- Change PIL to Pillow

* 2012-11-15 - Version 1.0
- Issue #2: Added ethernet support
- Issue #3: Added compatibility with libusb-1.0.1
- Issue #4: Fixed typo in escpos.py

* 2013-03-14 - Version 1.0.1
- Issue #8: Fixed set font
- Added QR support

* 2013-12-30 - Version 1.0.2
- Issue #5: Fixed vertical tab
- Issue #9: Fixed identation inconsistence

* 2013-02-23 - Version 1.0.3
- Issue #18: Added quad-area characters (Sent by syncman1x@gmail.com)
- Added exception for PIL import


escpos/escpos.py
#!/usr/bin/python
'''
@author: Manuel F Martinez <manpaz@bashlinux.com>
@organization: Bashlinux
@copyright: Copyright (c) 2012 Bashlinux
@license: GPL
'''

try:
    import Image
except ImportError:
    from PIL import Image

import qrcode
import time

from constants import *
from exceptions import *

class EscposIO(object):
    ''' ESC/POS Printer IO object'''
    def __init__(self, printer, autocut=True, autoclose=True):
        self.printer = printer
        self.params = {}
        self.autocut = autocut
        self.autoclose = autoclose


    def set(self, **kwargs):
        """
        :type bold:         bool
        :param bold:        set bold font
        :type underline:    [None, 1, 2]
        :param underline:   underline text
        :type size:         ['normal', '2w', '2h' or '2x']
        :param size:        Text size
        :type font:         ['a', 'b', 'c']
        :param font:        Font type
        :type align:        ['left', 'center', 'right']
        :param align:       Text position
        :type inverted:     boolean
        :param inverted:    White on black text
        :type color:        [1, 2]
        :param color:       Text color
        :rtype:             NoneType
        :returns:            None
        """

        self.params.update(kwargs)


    def writelines(self, text, **kwargs):
        params = dict(self.params)
        params.update(kwargs)

        if isinstance(text, unicode) or isinstance(text, str):
            lines = text.split('\n')
        elif isinstance(text, list) or isinstance(text, tuple):
            lines = text
        else:
            lines = ["{0}".format(text),]

        for line in lines:
            self.printer.set(**params)
            if isinstance(text, unicode):
                self.printer.text(u"{0}\n".format(line))
            else:
                self.printer.text("{0}\n".format(line))


    def close(self):
        self.printer.close()


    def __enter__(self, **kwargs):
        return self


    def __exit__(self, type, value, traceback):
        if not (type is not None and issubclass(type, Exception)):
            if self.autocut:
                self.printer.cut()

        if self.autoclose:
            self.close()



class Escpos(object):
    """ ESC/POS Printer object """
    device    = None
    _codepage  = None
    stored_args = {}


    def _check_image_size(self, size):
        """ Check and fix the size of the image to 32 bits """
        if size % 32 == 0:
            return (0, 0)
        else:
            image_border = 32 - (size % 32)
            if (image_border % 2) == 0:
                return (image_border / 2, image_border / 2)
            else:
                return (image_border / 2, (image_border / 2) + 1)


    def _print_image(self, line, size):
        """ Print formatted image """
        i = 0
        cont = 0
        buffer = ""

        self._raw(S_RASTER_N)
        buffer = "%02X%02X%02X%02X" % (((size[0]/size[1])/8), 0, size[1], 0)
        self._raw(buffer.decode('hex'))
        buffer = ""

        while i < len(line):
            hex_string = int(line[i:i+8],2)
            buffer += "%02X" % hex_string
            i += 8
            cont += 1
            if cont % 4 == 0:
                self._raw(buffer.decode("hex"))
                buffer = ""
                cont = 0


    def _convert_image(self, im):
        """ Parse image and prepare it to a printable format """
        pixels   = []
        pix_line = ""
        im_left  = ""
        im_right = ""
        switch   = 0
        img_size = [ 0, 0 ]


        if im.size[0] > 512:
            print  ("WARNING: Image is wider than 512 and could be truncated at print time ")
        if im.size[1] > 255:
            raise ImageSizeError()

        im_border = self._check_image_size(im.size[0])
        for i in range(im_border[0]):
            im_left += "0"
        for i in range(im_border[1]):
            im_right += "0"

        for y in range(im.size[1]):
            img_size[1] += 1
            pix_line += im_left
            img_size[0] += im_border[0]
            for x in range(im.size[0]):
                img_size[0] += 1
                RGB = im.getpixel((x, y))
                im_color = (RGB[0] + RGB[1] + RGB[2])
                im_pattern = "1X0"
                pattern_len = len(im_pattern)
                switch = (switch - 1 ) * (-1)
                for x in range(pattern_len):
                    if im_color <= (255 * 3 / pattern_len * (x+1)):
                        if im_pattern[x] == "X":
                            pix_line += "%d" % switch
                        else:
                            pix_line += im_pattern[x]
                        break
                    elif im_color > (255 * 3 / pattern_len * pattern_len) and im_color <= (255 * 3):
                        pix_line += im_pattern[-1]
                        break
            pix_line += im_right
            img_size[0] += im_border[1]

        self._print_image(pix_line, img_size)


    def image(self,path_img):
        """ Open image file """
        im_open = Image.open(path_img)
        im = im_open.convert("RGB")
        # Convert the RGB image in printable image
        self._convert_image(im)


    def qr(self, text, *args, **kwargs):
        """ Print QR Code for the provided string """
        qr_args = dict(
            version=4,
            box_size=4,
            border=1,
            error_correction=qrcode.ERROR_CORRECT_M
        )

        qr_args.update(kwargs)
        qr_code = qrcode.QRCode(**qr_args)

        qr_code.add_data(text)
        qr_code.make(fit=True)
        qr_img = qr_code.make_image()
        im = qr_img._img.convert("RGB")
        # Convert the RGB image in printable image
        self._convert_image(im)


    def barcode(self, code, bc, width, height, pos, font):
        """ Print Barcode """
        # Align Bar Code()
        self._raw(TXT_ALIGN_CT)
        # Height
        if height >=2 or height <=6:
            self._raw(BARCODE_HEIGHT)
        else:
            raise BarcodeSizeError()
        # Width
        if width >= 1 or width <=255:
            self._raw(BARCODE_WIDTH)
        else:
            raise BarcodeSizeError()
        # Font
        if font.upper() == "B":
            self._raw(BARCODE_FONT_B)
        else: # DEFAULT FONT: A
            self._raw(BARCODE_FONT_A)
        # Position
        if pos.upper() == "OFF":
            self._raw(BARCODE_TXT_OFF)
        elif pos.upper() == "BOTH":
            self._raw(BARCODE_TXT_BTH)
        elif pos.upper() == "ABOVE":
            self._raw(BARCODE_TXT_ABV)
        else:  # DEFAULT POSITION: BELOW
            self._raw(BARCODE_TXT_BLW)
        # Type
        if bc.upper() == "UPC-A":
            self._raw(BARCODE_UPC_A)
        elif bc.upper() == "UPC-E":
            self._raw(BARCODE_UPC_E)
        elif bc.upper() == "EAN13":
            self._raw(BARCODE_EAN13)
        elif bc.upper() == "EAN8":
            self._raw(BARCODE_EAN8)
        elif bc.upper() == "CODE39":
            self._raw(BARCODE_CODE39)
        elif bc.upper() == "ITF":
            self._raw(BARCODE_ITF)
        elif bc.upper() == "NW7":
            self._raw(BARCODE_NW7)
        else:
            raise BarcodeTypeError()
        # Print Code
        if code:
            self._raw(code)
        else:
            raise exception.BarcodeCodeError()


    def text(self, text):
        """ Print alpha-numeric text """
        if text:
            if self._codepage:
                self._raw(unicode(text).encode(self._codepage))
            else:
                self._raw(text)
        else:
            raise TextError()


    def set(self, codepage=None, **kwargs):
        """
        :type bold:         bool
        :param bold:        set bold font
        :type underline:    [None, 1, 2]
        :param underline:   underline text
        :type size:         ['normal', '2w', '2h' or '2x']
        :param size:        Text size
        :type font:         ['a', 'b', 'c']
        :param font:        Font type
        :type align:        ['left', 'center', 'right']
        :param align:       Text position
        :type inverted:     boolean
        :param inverted:    White on black text
        :type color:        [1, 2]
        :param color:       Text color
        :rtype:             NoneType
        :returns:            None
        """

        for key in kwargs.iterkeys():
            if not TEXT_STYLE.has_key(key):
                raise KeyError('Parameter {0} is wrong.'.format(key))

        for key, value in TEXT_STYLE.iteritems():
            if kwargs.has_key(key):
                cur = kwargs[key]
                if isinstance(cur, str) or isinstance(cur, unicode):
                    cur = cur.lower()

                if value.has_key(cur):
                    self._raw(value[cur])
                else:
                    raise AttributeError(
                        'Attribute {0} is wrong.'.format(cur)
                    )

        # Codepage
        self._codepage = codepage
        if codepage:
            self._raw(PAGE_CP_SET_COMMAND + chr(PAGE_CP_CODE[codepage]))

    def cut(self, mode='', postfix="\n\n\n\n\n"):
        """ Cut paper """
        # Fix the size between last line and cut
        # TODO: handle this with a line feed
        self._raw(postfix)
        if mode.upper() == "PART":
            self._raw(PAPER_PART_CUT)
        else: # DEFAULT MODE: FULL CUT
            self._raw(PAPER_FULL_CUT)


    def cashdraw(self, pin):
        """ Send pulse to kick the cash drawer """
        if pin == 2:
            self._raw(CD_KICK_2)
        elif pin == 5:
            self._raw(CD_KICK_5)
        else:
            raise CashDrawerError()


    def hw(self, hw):
        """ Hardware operations """
        if hw.upper() == "INIT":
            self._raw(HW_INIT)
        elif hw.upper() == "SELECT":
            self._raw(HW_SELECT)
        elif hw.upper() == "RESET":
            self._raw(HW_RESET)
        else: # DEFAULT: DOES NOTHING
            pass

        self._codepage = None


    def control(self, ctl):
        """ Feed control sequences """
        if ctl.upper() == "LF":
            self._raw(CTL_LF)
        elif ctl.upper() == "FF":
            self._raw(CTL_FF)
        elif ctl.upper() == "CR":
            self._raw(CTL_CR)
        elif ctl.upper() == "HT":
            self._raw(CTL_HT)
        elif ctl.upper() == "VT":
            self._raw(CTL_VT)

    def close(self):
        self._raw(RESET)
        self.__del__()


escpos/constants.py
""" ESC/POS Commands (Constants) """

# Feed control sequences
CTL_LF    = '\x0a'             # Print and line feed
CTL_FF    = '\x0c'             # Form feed
CTL_CR    = '\x0d'             # Carriage return
CTL_HT    = '\x09'             # Horizontal tab
CTL_VT    = '\x0b'             # Vertical tab
# Printer hardware
HW_INIT   = '\x1b\x40'         # Clear data in buffer and reset modes
HW_SELECT = '\x1b\x3d\x01'     # Printer select
HW_RESET  = '\x1b\x3f\x0a\x00' # Reset printer hardware
# Cash Drawer
CD_KICK_2 = '\x1b\x70\x00'     # Sends a pulse to pin 2 []
CD_KICK_5 = '\x1b\x70\x01'     # Sends a pulse to pin 5 []
# Paper
PAPER_FULL_CUT  = '\x1d\x56\x00' # Full cut paper
PAPER_PART_CUT  = '\x1d\x56\x01' # Partial cut paper
# Text format

BARCODE_TXT_OFF = '\x1d\x48\x00' # HRI barcode chars OFF
BARCODE_TXT_ABV = '\x1d\x48\x01' # HRI barcode chars above
BARCODE_TXT_BLW = '\x1d\x48\x02' # HRI barcode chars below
BARCODE_TXT_BTH = '\x1d\x48\x03' # HRI barcode chars both above and below
BARCODE_FONT_A  = '\x1d\x66\x00' # Font type A for HRI barcode chars
BARCODE_FONT_B  = '\x1d\x66\x01' # Font type B for HRI barcode chars
BARCODE_HEIGHT  = '\x1d\x68\x64' # Barcode Height [1-255]
BARCODE_WIDTH   = '\x1d\x77\x03' # Barcode Width  [2-6]
BARCODE_UPC_A   = '\x1d\x6b\x00' # Barcode type UPC-A
BARCODE_UPC_E   = '\x1d\x6b\x01' # Barcode type UPC-E
BARCODE_EAN13   = '\x1d\x6b\x02' # Barcode type EAN13
BARCODE_EAN8    = '\x1d\x6b\x03' # Barcode type EAN8
BARCODE_CODE39  = '\x1d\x6b\x04' # Barcode type CODE39
BARCODE_ITF     = '\x1d\x6b\x05' # Barcode type ITF
BARCODE_NW7     = '\x1d\x6b\x06' # Barcode type NW7

# Image format
S_RASTER_N      = '\x1d\x76\x30\x00' # Set raster image normal size
S_RASTER_2W     = '\x1d\x76\x30\x01' # Set raster image double width
S_RASTER_2H     = '\x1d\x76\x30\x02' # Set raster image double height
S_RASTER_Q      = '\x1d\x76\x30\x03' # Set raster image quadruple

RESET           = '\x1b\x40'

TEXT_STYLE = {
    'bold': {
        0: '\x1b\x45\x00',          # Bold font OFF
        1: '\x1b\x45\x01',          # Bold font ON
    },
    'underline': {
        None: '\x1b\x2d\x00',       # Underline font OFF
        1: '\x1b\x2d\x01',          # Underline font 1-dot ON
        2: '\x1b\x2d\x02',          # Underline font 2-dot ON
    },
    'size': {
        'normal': '\x1b\x21\x00',   # Normal text
        '2h': '\x1b\x21\x10',       # Double height text
        '2w': '\x1b\x21\x20',       # Double width text
        '2x': '\x1b\x21\x30',       # Quad area text
    },
    'font': {
        'a': '\x1b\x4d\x00',        # Font type A
        'b': '\x1b\x4d\x01',        # Font type B
        'c': '\x1b\x4d\x02',        # Font type C (may not support)
    },
    'align': {
        'left': '\x1b\x61\x00',     # Left justification
        'right': '\x1b\x61\x02',    # Right justification
        'center': '\x1b\x61\x01',   # Centering
    },
    'inverted': {
        False: '\x1d\x42\x00',      # Inverted mode ON
        True: '\x1d\x42\x01',       # Inverted mode OFF
    },
    'color': {
        1: '\x1b\x72\x00',          # Select 1st printing color
        2: '\x1b\x72\x00',          # Select 2nd printing color
    }
}

PAGE_CP_SET_COMMAND = '\x1b\x74'
PAGE_CP_CODE = {
    'cp437'       : 0,
    # 'katakana'  : 1,
    'cp850'       : 2,
    'cp860'       : 3,
    'cp863'       : 4,
    'cp865'       : 5,
    'cp1251'      : 6,
    'cp866'       : 7,
    'mac_cyrillic': 8,
    'cp775'       : 9,
    'cp1253'      : 10,
    'cp737'       : 11,
    'cp857'       : 12,
    'iso8859_9'   : 13,
    'cp864'       : 14,
    'cp862'       : 15,
    'iso8859_2'   : 16,
    'cp1253'      : 17,
    'cp1250'      : 18,
    'cp858'      : 19,
    'cp1254'      : 20,
    # 'TIS_14'    : 21,
    # 'TIS_17'    : 22,
    # 'TIS_11'    : 23,
    'cp737'       : 24,
    'cp1257'      : 25,
    'cp847'       : 26,
    # 'cp720'     : 27,
    'cp885'       : 28,
    'cp857'       : 29,
    'cp1250'      : 30,
    'cp775'       : 31,
    'cp1254'      : 32,
    # ''          : 33,
    'cp1256'      : 34,
    'cp1258'      : 35,
    'iso8859_2'   : 36,
    'iso8859_3'   : 37,
    'iso8859_4'   : 38,
    'iso8859_5'   : 39,
    'iso8859_6'   : 40,
    'iso8859_7'   : 41,
    'iso8859_8'   : 42,
    'iso8859_9'   : 43,
    'iso8859_15'  : 44,
    # '???'       : 45,
    'cp856'       : 46,
    'cp874'       : 47,
}
